tools/snapshot/snapshot.py

```diff
@@ -0,0 +1,468 @@
+    log.debug(f"adding ssh key {ssh_key['path']} to ssh agent")
+    if ssh_key['passphrase'] is None:
+        subprocess.check_call(["ssh-add", ssh_key['path']], env=os.environ)
+    else:
+        proc = pexpect.spawn("ssh-add", [ssh_key['path']], env=os.environ)
+        proc.logfile = sys.stdout
+        proc.sendline(ssh_key['passphrase'])
+        proc.expect(pexpect.EOF)
+        log.debug(proc.before.decode())
+
+
+def _ssh_cmd(user, ip, command):
+    """
+    Internal SSH command. Assumes ssh agent has been initialized.
+
+    Returns the output of the SSH command.
+    Raises subprocess.CalledProcessError if ssh call errored.
+    """
+    cmd = ["ssh", f"{user}@{ip}"] if ssh_key['use_existing_agent'] else ["ssh", "-i", ssh_key["path"], f"{user}@{ip}"]
+    cmd.append(command)
+    return subprocess.check_output(cmd, env=os.environ).decode()
+
+
+def _ssh_script(user, ip, bash_script_path):
+    """
+    Internal SSH command. Assumes ssh agent has been initialized.
+
+    Returns the output of the SSH command.
+    Raises subprocess.CalledProcessError if ssh call errored.
+    """
+    cmd = ["ssh", f"{user}@{ip}"] if ssh_key['use_existing_agent'] else ["ssh", "-i", ssh_key["path"], f"{user}@{ip}"]
+    cmd.append('bash -s')
+    with open(bash_script_path, 'rb') as f:
+        return subprocess.check_output(cmd, env=os.environ, stdin=f).decode()
+
+
+def load_config(config_path):
+    """
+    Load the given config.
+
+    Raises FileNotFoundError, KeyError, or json.decoder.JSONDecodeError
+    """
+    with open(config_path, 'r', encoding="utf-8") as f:
+        config = json.load(f)
+    if {'machines', 'ssh_key', 'rsync', 'condition'} != set(config.keys()):
+        raise KeyError(f"config keys: {config.keys()} do not contain 'machines', 'ssh_key', 'condition' or 'rsync'.")
+    log.debug(f"config: {json.dumps(config, indent=2)}")
+    machines.clear()
+    rsync.clear()
+    ssh_key.clear()
+    condition.clear()
+    machines.extend(config['machines'])
+    rsync.update(config['rsync'])
+    ssh_key.update(config['ssh_key'])
+    condition.update(config['condition'])
+    _init_ssh_agent()
+
+
+def _is_dns_node(machine, sharding_structure):
+    """
+    Internal function to check if given machine is a DNS node.
+
+    Note the assumptions in the format of the HTTP endpoint given by `sharding_structure`.
+    """
+    for structure in sharding_structure:
+        if int(structure['shardID']) == int(machine['shard']):
+            http = structure['http'].strip()
+            assert http.startswith('https://api.'), f"unknown endpoint format from machine {machine['ip']}"
+            for ip in dns.resolver.query(http.replace('https://api.', '')):
+                if machine['ip'] == str(ip):
+                    return True
+            return False
+    raise RuntimeError(f"unknown shard for network (sharding structure "
+                       f"endpoint not found for shard {machine['shard']})")
+
+
+def sanity_check():
```

Comment:
good check! can you add some comments here?

Reply:
sure will add comments